File: Week 29/cpp/rate_limiting.cpp
/*
 * WEEK 29 - C++ ADVANCED TOPICS
 * Topic: Rate Limiting - Token Bucket, Leaky Bucket, Sliding Window
 * File: rate_limiting.cpp
 *
 * CONCEPT:
 *   Rate limiters cap how often an action may occur:
 *     - Token Bucket: capacity tokens refilling at rate; consume per req.
 *     - Leaky Bucket: queue draining at constant rate; overflow rejected.
 *     - Sliding Window Log: timestamps in last T; allow if count < limit.
 *
 * KEY POINTS:
 *   - Token bucket allows bursts up to capacity; enforces average rate.
 *   - Leaky bucket smooths bursts into a steady stream.
 *   - Sliding log gives exact answers but O(limit) memory.
 *
 * ALGORITHM / APPROACH:
 *   TOKEN:  refill = elapsed * rate; cap at capacity; consume if >=1.
 *   LEAKY:  drain by elapsed * rate; admit if water < cap; water += 1.
 *   SLIDING: pop timestamps older than now-T; allow if size < max; push now.
 *
 * C++-SPECIFIC NOTES:
 *   - std::chrono::steady_clock for monotonic clock.
 *   - std::deque<TimePoint> for the sliding window log.
 *
 * DRY RUN / EXAMPLE:
 *   TokenBucket(5, 2 tok/s) -> first 5 requests allowed, 6th rejected;
 *   500 ms later one more allowed.
 *
 * COMPLEXITY:
 *   Token/Leaky O(1); Sliding O(window). Space O(window).
 */

#include <iostream>
#include <chrono>
#include <deque>
#include <thread>

using namespace std;
using namespace std::chrono;

class TokenBucket {
    int cap;
    double rate;
    double tokens;
    steady_clock::time_point last;
public:
    TokenBucket(int c, double r) : cap(c), rate(r), tokens(c), last(steady_clock::now()) {}
    bool allow() {
        auto now = steady_clock::now();
        double elapsed = duration<double>(now - last).count();
        tokens = min((double)cap, tokens + elapsed * rate);
        last = now;
        if (tokens >= 1.0) { tokens -= 1.0; return true; }
        return false;
    }
};

class LeakyBucket {
    int cap;
    double rate;
    double water;
    steady_clock::time_point last;
public:
    LeakyBucket(int c, double r) : cap(c), rate(r), water(0), last(steady_clock::now()) {}
    bool allow() {
        auto now = steady_clock::now();
        double elapsed = duration<double>(now - last).count();
        water = max(0.0, water - elapsed * rate);
        last = now;
        if (water < cap) { water += 1.0; return true; }
        return false;
    }
};

class SlidingWindowLog {
    int maxReq;
    duration<double> window;
    deque<steady_clock::time_point> log;
public:
    SlidingWindowLog(int n, double seconds) : maxReq(n), window(seconds) {}
    bool allow() {
        auto now = steady_clock::now();
        while (!log.empty() && log.front() <= now - window) log.pop_front();
        if ((int)log.size() < maxReq) { log.push_back(now); return true; }
        return false;
    }
};

int main() {
    TokenBucket tb(5, 2.0);
    for (int i = 1; i <= 7; ++i)
        cout << "  TokenBucket req " << i << ": " << (tb.allow() ? "ALLOWED" : "REJECTED") << "\n";

    cout << "\n";
    LeakyBucket lb(3, 5.0);
    for (int i = 1; i <= 5; ++i)
        cout << "  LeakyBucket req " << i << ": " << (lb.allow() ? "ALLOWED" : "REJECTED") << "\n";

    cout << "\n";
    SlidingWindowLog sw(3, 0.05);
    for (int i = 1; i <= 5; ++i)
        cout << "  SlidingWindow req " << i << ": " << (sw.allow() ? "ALLOWED" : "REJECTED") << "\n";
    return 0;
}

/*
 * NOTES
 * -----
 * Differences from Java:
 *   - We add leaky bucket and sliding-window log for spec completeness.
 *   - std::chrono::steady_clock for the monotonic clock.
 *   - duration<double> in seconds keeps arithmetic clean.
 */

File: Week 29/rust/rate_limiting.rs
// WEEK 29 - RUST ADVANCED TOPICS
// Topic: Rate Limiting - Token Bucket, Leaky Bucket, Sliding Window Log
// File: rate_limiting.rs
//
// CONCEPT:
//   Cap how often an action can occur:
//     - Token Bucket: capacity tokens refilled at rate; consume per req.
//     - Leaky Bucket: queue draining at constant rate; overflow rejected.
//     - Sliding Window Log: timestamps in last T; allow if count < limit.
//
// KEY POINTS:
//   - Token bucket allows bursts up to capacity; enforces average rate.
//   - Leaky bucket smooths bursts into a steady stream.
//   - Sliding log gives exact answers but O(limit) memory.
//
// ALGORITHM / APPROACH:
//   TOKEN:  refill = elapsed * rate; cap at capacity; consume if >= 1.
//   LEAKY:  water -= elapsed * rate (floored 0); if < cap admit + water+=1.
//   SLIDING: pop ts older than now-T; allow if len < limit; push now.
//
// RUST-SPECIFIC NOTES:
//   - std::time::Instant for monotonic clock.
//   - VecDeque<Instant> for sliding log (O(1) front pop).
//
// DRY RUN / EXAMPLE:
//   TokenBucket(5, 2.0) -> 5 reqs allowed back-to-back, 6th rejected;
//   after 500 ms one more allowed (one token refilled).
//
// COMPLEXITY:
//   Token/Leaky O(1) per op; Sliding O(window). Space O(window).

use std::collections::VecDeque;
use std::thread;
use std::time::{Duration, Instant};

pub struct TokenBucket { cap: f64, rate: f64, tokens: f64, last: Instant }
impl TokenBucket {
    pub fn new(cap: u32, rate_per_sec: f64) -> Self {
        Self { cap: cap as f64, rate: rate_per_sec, tokens: cap as f64, last: Instant::now() }
    }
    pub fn allow(&mut self) -> bool {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last).as_secs_f64();
        self.tokens = (self.tokens + elapsed * self.rate).min(self.cap);
        self.last = now;
        if self.tokens >= 1.0 { self.tokens -= 1.0; true } else { false }
    }
}

pub struct LeakyBucket { cap: f64, rate: f64, water: f64, last: Instant }
impl LeakyBucket {
    pub fn new(cap: u32, rate_per_sec: f64) -> Self {
        Self { cap: cap as f64, rate: rate_per_sec, water: 0.0, last: Instant::now() }
    }
    pub fn allow(&mut self) -> bool {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last).as_secs_f64();
        self.water = (self.water - elapsed * self.rate).max(0.0);
        self.last = now;
        if self.water < self.cap { self.water += 1.0; true } else { false }
    }
}

pub struct SlidingWindowLog { max: usize, window: Duration, log: VecDeque<Instant> }
impl SlidingWindowLog {
    pub fn new(max: usize, window_seconds: f64) -> Self {
        Self { max, window: Duration::from_secs_f64(window_seconds), log: VecDeque::new() }
    }
    pub fn allow(&mut self) -> bool {
        let now = Instant::now();
        while let Some(&front) = self.log.front() {
            if now.duration_since(front) > self.window { self.log.pop_front(); }
            else { break; }
        }
        if self.log.len() < self.max { self.log.push_back(now); true } else { false }
    }
}

fn main() {
    let mut tb = TokenBucket::new(5, 2.0);
    for i in 1..=7 {
        println!("  TokenBucket req {}: {}", i,
                 if tb.allow() { "ALLOWED" } else { "REJECTED" });
    }
    println!();
    let mut lb = LeakyBucket::new(3, 5.0);
    for i in 1..=5 {
        println!("  LeakyBucket req {}: {}", i,
                 if lb.allow() { "ALLOWED" } else { "REJECTED" });
    }
    println!();
    let mut sw = SlidingWindowLog::new(3, 0.05);
    for i in 1..=5 {
        println!("  SlidingWindow req {}: {}", i,
                 if sw.allow() { "ALLOWED" } else { "REJECTED" });
        // No sleep — burst test
    }
    // Demonstrate expiry
    thread::sleep(Duration::from_millis(60));
    println!("After 60 ms wait:");
    println!("  SlidingWindow req 6: {}", if sw.allow() { "ALLOWED" } else { "REJECTED" });
}

// NOTES
// -----
// Differences from Java:
//   * Adds leaky-bucket and sliding-window log for spec completeness.
//   * std::time::Instant + duration_since for monotonic clocks.
//   * VecDeque<Instant> is the idiomatic sliding-window buffer.
